components/script/script_module.rs

```diff
@@ -0,0 +1,1365 @@
+        let rejected_url = module_url.clone();
+
+        let handler = PromiseNativeHandler::new(
+            &owner.global(),
+            Some(ModuleHandler::new(Box::new(
+                task!(fetched_resolve: move || {
+                    println!("fetched: {}", resolved_url.clone());
+                    resolve_this.finish_module_load(Some(resolved_url));
+                }),
+            ))),
+            Some(ModuleHandler::new(Box::new(
+                task!(failure_reject: move || {
+                    println!("failed");
+                    reject_this.finish_module_load(Some(rejected_url));
+                }),
+            ))),
+        );
+
+        let _compartment = enter_realm(&*owner.global());
+        AlreadyInCompartment::assert(&*owner.global());
+        let _ais = AutoIncumbentScript::new(&*owner.global());
+
+        let promise = promise.as_ref().unwrap();
+
+        promise.append_native_handler(&handler);
+    }
+}
+
+#[derive(Clone, Copy, Debug, JSTraceable, PartialEq, PartialOrd)]
+pub enum ModuleStatus {
+    Initial,
+    Fetching,
+    FetchFailed,
+    FetchingDescendants,
+    Finished,
+}
+
+impl ModuleTree {
+    #[allow(unsafe_code)]
+    /// https://html.spec.whatwg.org/multipage/#creating-a-module-script
+    /// Step 7-11.
+    fn compile_module_script(
+        &self,
+        global: &GlobalScope,
+        module_script_text: DOMString,
+        url: ServoUrl,
+    ) -> Result<ModuleObject, ModuleError> {
+        let module: Vec<u16> = module_script_text.encode_utf16().collect();
+
+        let url_cstr = ffi::CString::new(url.as_str().as_bytes()).unwrap();
+
+        let _ac = JSAutoRealm::new(*global.get_cx(), *global.reflector().get_jsobject());
+
+        let compile_options = CompileOptionsWrapper::new(*global.get_cx(), url_cstr.as_ptr(), 1);
+
+        rooted!(in(*global.get_cx()) let mut module_script = ptr::null_mut::<JSObject>());
+
+        let mut source = get_source_text(module);
+
+        unsafe {
+            if !CompileModule(
+                *global.get_cx(),
+                compile_options.ptr,
+                &mut source,
+                &mut module_script.handle_mut(),
+            ) {
+                println!("fail to compile module script of {}", url);
+
+                rooted!(in(*global.get_cx()) let mut exception = UndefinedValue());
+                assert!(JS_GetPendingException(
+                    *global.get_cx(),
+                    &mut exception.handle_mut()
+                ));
+                JS_ClearPendingException(*global.get_cx());
+
+                return Err(ModuleError::RawException(RootedTraceableBox::from_box(
+                    Heap::boxed(exception.get()),
+                )));
+            }
+        }
+
+        println!("module script of {} compile done", url);
+
+        self.resolve_requested_module_specifiers(
+            &global,
+            module_script.handle().into_handle(),
+            global.api_base_url().clone(),
```

Comment:
The spec says this should be the [script's base URL](https://html.spec.whatwg.org/multipage/webappapis.html#concept-script-base-url), rather than the document's base URL. This is the cause of many assertion errors that I encountered when trying out various three.js content, which requires ../../../build/three.module.js from http://rawcdn.githack.com/mrdoob/three.js/master/examples/jsm/controls/OrbitControls.js but ends up resolving against http://rawcdn.githack.com/mrdoob/three.js/master/examples/webgl_animation_cloth.html, http://rawcdn.githack.com/mrdoob/build/three.module.js which gives a 404.

Reply:
Well, it's half of the reason for the assertions. The other half is the `TODO: Step 3` in HostResolveImportedModule, which is the other caller of resolve_module_specifier that needs to pass the script's base URL when appropriate.

Reply:
> The spec says this should be the script's base URL, rather than the document's base URL.

Thanks for catching this 🙏

> The other half is the TODO: Step 3 in HostResolveImportedModule, which is the other caller of resolve_module_specifier that needs to pass the script's base URL when appropriate.

Thanks for the patch at https://github.com/servo/servo/pull/23545#issuecomment-556063220. I've picked the changes into this PR.

I wonder while we've handled the `SetModulePrivate`, we might also want to handle the module private hook in this PR as well 🤔

Gecko registers two functions for `SetScriptPrivateReferenceHooks`; one for incrementing the reference count and the other for releasing the reference count. But, it isn't clear to me how we can handle the same thing in Servo 🤔(I can imagine we might need to call `deref()` in Rust for the release one?) Please correct me if I misunderstood anything! 🙇 Thanks!

https://searchfox.org/mozilla-central/rev/652014ca1183c56bc5f04daf01af180d4e50a91c/dom/script/ScriptLoader.cpp#1043-1044